internal/connection/connection.go

```diff
@@ -29,23 +31,68 @@ import (
 	"google.golang.org/grpc/credentials/insecure"
 )
 
+const (
+	// The duration after which a gRPC connection is closed due to inactivity
+	IDLE_TIMEOUT = time.Minute * 5
+)
+
 // Connection struct consists of to NodeID, DriverName, Capabilities for the controller
 // to pick sidecar connection and grpc Client to connect to the sidecar.
 type Connection struct {
+	sync.Mutex
+
 	Client       *grpc.ClientConn
 	Capabilities []*identity.Capability
 	Namespace    string
 	Name         string
 	NodeID       string
 	DriverName   string
 	Timeout      time.Duration
+
+	// Holds the internal state of the connection
+	connected  bool
+	enableAuth bool
+	endpoint   string
+	driverName string
+	podName    string
+
+	// Used to cancel any existing timers in case of re-connects
+	cancelIdle context.CancelFunc
+	// Holds the last access time of the gRPC Connection
+	// It is tracked and updated by the `accessTimeInterceptor`
+	lastAccessTime atomic.Int64
 }
 
-// NewConnection establishes connection with sidecar, fetches capability and returns Connection object
-// filled with required information.
-func NewConnection(ctx context.Context, endpoint, nodeID, driverName, namespace, podName string, enableAuth bool) (*Connection, error) {
-	var opts []grpc.DialOption
-	if enableAuth {
+// accessTimeInterceptor is an unary interceptor which updates the lastAccessTime
+// on `Connection` struct to `time.Now()` on each RPC call
+func accessTimeInterceptor(conn *Connection) grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req, reply any, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		conn.lastAccessTime.Store(time.Now().UnixNano())
+
+		return invoker(ctx, method, req, reply, cc, opts...)
+	}
+}
+
+// Connect creates a new grpc.ClientConn object and sets it as the
+// client property on Connection struct. If a connection is already
+// connected, it is reused as is. It also spawns a go routine to tear
+// down the connection if it has been idling for a specified threshold.
+//
+// In cases where a new connection is created from the scratch Connect
+// also calls fetchCapabilities on the connection object.
+func (c *Connection) Connect() error {
+	c.Lock()
+	defer c.Unlock()
+
+	// Return early if already connected, the connection will be reused
+	if c.connected {
```

Comment:
maybe update `conn.lastAccessTime.Store(time.Now().UnixNano())` here?

It might be cleaner to have that as a function like the below, so that storing a new timestamp is done only in one place.
```go
func (c *Connection) touch() {
    c.lastAccessTime.Store(time.Now().UnixNano())
}
```

Reply:
Is that really necessary here? The interceptor (middleware) will update it on every request it processes? After returning early if already connected, ideally there will be subsequent requests, which in turn will update the last access time.

Reply:
The ticker doesn't send an event until the first tick which is in 30 seconds.